### PR TITLE
[Diagnostics] Add diagnostic for non-objc functions declared as open inside extension

### PR DESCRIPTION
**Pull request by mininny**
The following code snippet:
```
public class C { }
extension C {
  open func f() { }
}
```
compiles without a diagnostic today. Superficially, it might seem like you can override f from a subclass of C. However, that's not the case; you can only overload it. Being able to override methods declared in extensions would require a mechanism like objc_msgSend, which is why such overloading only works when @objc is present, there is no Swift equivalent for pure Swift classes.

This means that the open is misleading to a reader; we should issue a diagnostic with a fix-it to replace this with public instead in such situations.

Resolves [SR-15096](https://bugs.swift.org/browse/SR-15096)

### CODE REVIEW

File: lib/Sema/TypeCheckAttr.cpp
Comment (varungandhi-apple) [inline lib/Sema/TypeCheckAttr.cpp:802]: Either there is an extra space here or there is a missing space in the `"public"` below this. But it seems like they should be consistent.

Comment (varungandhi-apple) [reply lib/Sema/TypeCheckAttr.cpp:802]: Also, related to my comment in the other PR; I think this should apply to properties too and not just methods.

File: include/swift/AST/DiagnosticsSema.def
Comment (varungandhi-apple) [inline include/swift/AST/DiagnosticsSema.def:1508]: This is potentially a good opportunity to briefly explain why `open` cannot be used. Say something like:

```
"non-@objc %select{methods|properties}0 in extensions cannot be overriden; use 'public' instead"
```

**Comment by varungandhi-apple** [comment]
@swift-ci test source compatibility

File: lib/Sema/TypeCheckAttr.cpp
Comment (mininny) [reply lib/Sema/TypeCheckAttr.cpp]: Thanks! I removed the extra space and applied this to properties as well :)

File: lib/Sema/TypeCheckAttr.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/TypeCheckAttr.cpp:801]: Nit: This seems like missing clang format

Comment (mininny) [reply lib/Sema/TypeCheckAttr.cpp]: Fixed :)

**Comment by YOCKOW** [comment]
I have a question just for my curiosity: *Why is this an error rather than a warning?*

I understand `func f()` should not be `open`, but the sample code you describe is compiled successfully so far and `f` itself can be called without any error. 
Why don't we take the backward compatibility into account in this case?

**Comment by mininny** [comment]
> That was one of the concerns I had, which is why I ran the source compat suite testing. I'm not sure what happened with the results, but I can re-run it.
> 
> @swift-ci test source compatibility

It seems like the source compat suite failed due to this change. Should I change the diagnostic to a warning instead of an error?

**Comment by LucianoPAlmeida** [comment]
> It seems like the source compat suite failed due to this change. Should I change the diagnostic to a warning instead of an error?

Looking into the source compatibility logs, even if this is all it seems like we have to do an extra check also to not emit when compiling module interface declarations (at least for objective C modules) in the example of source compat AppKit was one that broke...

**Comment by mininny** [comment]
> > It seems like the source compat suite failed due to this change. Should I change the diagnostic to a warning instead of an error?
> 
> Looking into the source compatibility logs, even if this is all it seems like we have to do an extra check also to not emit when compiling module interface declarations (at least for objective C modules) in the example of source compat AppKit was one that broke...

I see 👍 I lowered the error to warning, and added a check for module interface(I'm not sure if I'm correctly doing what you said). If this is correct, can someone please re-run the source compat tests? :)

**Comment by LucianoPAlmeida** [comment]
> I see 👍 I lowered the error to warning, and added a check for module interface(I'm not sure if I'm correctly doing what you said). 

It does make sense to me not to emit this diagnostic(warning or error) for module interface declarations, but I don't know enough about how objc modules swift interface files are generated, so maybe there is something I'm missing so let's wait for @varungandhi-apple thoughts on this =]

About the source compatibility, not sure it would worth run it with this changed to warning because it is not going to fail because the warning would not break compilation in any scenario...

**Comment by varungandhi-apple** [comment]
> but I don't know enough about how objc modules swift interface files are generated

Are you referring to the ones generated by SourceKit in Xcode or the swiftinterfaces for modules which have overlays (such as AppKit)? For the former, I don't think this warning will be emitted since those are synthesized. For the latter, the presence of the overlay means that the swiftinterface is generated when the overlay is compiled.

Becca pointed out that all warnings in swiftinterfaces are suppressed by default, so we shouldn't need special-casing here.

https://github.com/apple/swift/blob/main/lib/Frontend/ModuleInterfaceLoader.cpp#L1310-L1313

**Comment by LucianoPAlmeida** [comment]
> Are you referring to the ones generated by SourceKit in Xcode or the swiftinterfaces for modules which have overlays (such as AppKit)? For the former, I don't think this warning will be emitted since those are synthesized. For the latter, the presence of the overlay means that the swiftinterface is generated when the overlay is compiled.
>
> Becca pointed out that all warnings in swiftinterfaces are suppressed by default, so we shouldn't need special-casing here.

I was referring to the latter, good to know we don't need special case for warnings, thank you for the info!

**Comment by mininny** [comment]
> Are you referring to the ones generated by SourceKit in Xcode or the swiftinterfaces for modules which have overlays (such as AppKit)? For the former, I don't think this warning will be emitted since those are synthesized. For the latter, the presence of the overlay means that the swiftinterface is generated when the overlay is compiled.
> 
> Becca pointed out that all warnings in swiftinterfaces are suppressed by default, so we shouldn't need special-casing here.
> 
> https://github.com/apple/swift/blob/main/lib/Frontend/ModuleInterfaceLoader.cpp#L1310-L1313

Thank you for the clarification! I only lowered the error to a warning, and removed the check for module interface since it seems like it's not necessary :) 
Should this be then good to go?

File: include/swift/AST/DiagnosticsSema.def
Comment (varungandhi-apple) [inline include/swift/AST/DiagnosticsSema.def:1507]:
```suggestion
WARNING(access_control_non_objc_open_member,none,
```

**Review by varungandhi-apple — APPROVED** [review]

File: include/swift/AST/DiagnosticsSema.def
Comment (mininny) [reply include/swift/AST/DiagnosticsSema.def]: Fixed! :)

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test and merge

**Comment by mininny** [comment]
I didn't realize that this change affected existing test cases 😅 Could you please re-run the tests?

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test and merge

File: test/Compatibility/accessibility.swift
Comment (LucianoPAlmeida) [inline test/Compatibility/accessibility.swift:193]: That just make me wonder if we should say `use '${parent_decl_modifier}'` in this case `use 'internal'` instead, or even if we should emit this warning if the extension decl is less than 'public' ... @varungandhi-apple?

Comment (varungandhi-apple) [reply test/Compatibility/accessibility.swift:193]: That's a good point. In this case, I think we should skip the `non-@objc` warning because the other warning is more accurate.

Comment (mininny) [reply test/Compatibility/accessibility.swift]: Makes sense! 👍 I fixed the code and test to skip this warning in such cases.

File: test/attr/open_objc.swift
Comment (varungandhi-apple) [inline test/attr/open_objc.swift:8]: Nit: Please change the comment to `No warning` or `No warning: <reason>`. I think the "error" is an artifact of this being an error instead of a warning earlier.

Comment (mininny) [reply test/attr/open_objc.swift]: Fixed! :)

Comment (mininny) [reply test/attr/open_objc.swift]: Fixed! :)

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test Linux

**Comment by YOCKOW** [comment]
@swift-ci Please smoke test Linux platform

**Comment by LucianoPAlmeida** [comment]
@swift-ci smoke test Linux

File: include/swift/AST/DiagnosticsSema.def
Comment (xwu) [inline include/swift/AST/DiagnosticsSema.def:1508]: There is a typo (also, suggest using single quotation marks around `@objc` for code voice):

```suggestion
      "non-'@objc' %select{properties|methods}0 in extensions cannot be overridden; use 'public' instead",
```

**Review by xwu — CHANGES_REQUESTED** [review]
This is a very useful warning, I think. Just an issue with spelling to be fixed.

**Comment by xwu** [comment]
@swift-ci test and merge

**Comment by mininny** [comment]
> @swift-ci test and merge

The linux platform test is failing again with some error from corelibs-foundation, and I also see a lot of warnings for non-@objc open methods in extension from corelibs-foundation as well. 
The warning seems to be different from the build error itself, but I assume this may have caused the build to fail?

**Comment by xwu** [comment]
Looks unrelated to this PR.

**Comment by varungandhi-apple** [comment]
I think Minhyuk's evaluation is on the right track. This is the [declaration](https://github.com/apple/swift-corelibs-foundation/blob/01efe01a792e0070da7c42f54089b7a9dbb35bb4/Sources/Foundation/FileHandle.swift#L727) in swift-corelibs-foundation:

```
    open class var standardInput: FileHandle {
        return _stdinFileHandle
    }
```

ObjC-interop is not supported on Linux, so this declaration is non-`@objc` only on Linux. So this warning is issued:

```
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-main/swift-corelibs-foundation/Sources/Foundation/FileHandle.swift:727:5: warning: non-'@objc' properties in extensions cannot be overridden; use 'public' instead
    open class var standardInput: FileHandle {
    ^~~~
    public
```

When the warning is issued, the attribute is set to invalid. Later, something else tries to access the attribute and instead of treating it as open/public, it gets treated as internal for some reason (even thought it is marked invalid). The attribute being treated as internal triggers an error.

Open to other ideas as well, but I think it would make sense to:
1. At least take a look at why an invalid attribute is being treated as internal; it seems surprising to me that this is happening.
2. Remove the part marking the attribute as invalid and only issue the diagnostic.
3. Fix the code in swift-corelibs-foundation to move open properties inside the definition of `FileHandle`, instead of inside the extension.

**Comment by mininny** [comment]
Thanks for summing this up! I will look into the next steps you mentioned :) 

> 3\. Fix the code in swift-corelibs-foundation to move open properties inside the definition of `FileHandle`, instead of inside the extension.

I have a (maybe stupid) question about this.  

Looking at the CI, there seem to be many other cases in corelibs-foundation where this warning is occurring(i.e. in NSURL.swift) and the same error could occur(open→internal). I understand that this is only problematic in Linux, but since the syntax is technically incorrect and possibly misleading, shouldn't they all be fixed as well?

**Comment by varungandhi-apple** [comment]
> Looking at the CI, there seem to be many other cases in corelibs-foundation where this warning is occurring(i.e. in NSURL.swift) and the same error could occur(open→internal). I understand that this is only problematic in Linux, but since the syntax is technically incorrect and possibly misleading, shouldn't they all be fixed as well?

That's right. I was being fast and loose, didn't take a look at all warnings. But if there are similar warnings for other types too, it makes sense to update those as well. Also, this step can happen after landing the compiler-side change, since we're only adding a warning here, it doesn't have to land before the compiler change.

**Comment by xwu** [comment]
IMHO, the key issue that needs to be addressed revealed by the testing is that any declarations should continue to be `open` for the purposes of mangling, access control, etc.,  until the user fixes their code.

**Comment by mininny** [comment]
I did a rough search on this, and what I think is causing the attributes to be ignored is that `DeclAttributes::getAttribute/s` filter out invalid attributes, causing the invalid ACL attributes to be ignored as well, always resulting in a defaulted `internal` level. 
```
  template <typename ATTR>
  ATTR *getAttribute(bool AllowInvalid = false) {
    for (auto Attr : *this)
      if (auto *SpecificAttr = dyn_cast<ATTR>(Attr))
        if (SpecificAttr->isValid() || AllowInvalid)
          return SpecificAttr;
    return nullptr;
  }
```

If the attributes are indeed being applied to the context using the above function, I think there can be two ways to solve this: 
1. Find callers of above function, and mark `AllowInvalid` parameter as true.
2. Override the `isValid` and `isInvalid` in `AbstractAccessControlAttr` to always return as being valid. (Probably dangerous, I assume)

Come to think of it, I think similar issues can maybe arise in other types of attributes as well, such as `AvailableAttr` where the availability of a declaration should be visible from the call site.

**Comment by xwu** [comment]
> If the attributes are indeed being applied to the context using the above function, I think there can be two ways to solve this:

Why mark these uses of ˋopen` as invalid at all?

**Comment by varungandhi-apple** [comment]
> Why mark these uses of `open` as invalid at all?

One potential reason to do this would be if the `open` pessimizes optimizations downstream because the method can be overriden. That said, I don't know if that's actually the case here.

I think it is probably best to move forward here by not marking the `open` as invalid, and file a separate JIRA to track whether that ought to be changed (and link the discussion from here in that JIRA).

**Comment by mininny** [comment]
It's been several months, but I finally got the time to look at this again :) 
As discussed above, I removed the `invalid` mark from the attributes so that we can show this diagnostic first.

After this gets merged, I will file a separate JIRA for any further actions on the invalid attributes. 

Can someone check this and run the tests please?

**Comment by xwu** [comment]
@swift-ci test

**Comment by mininny** [comment]
It seems like some tests with DocC are failing. Can you run them again? 🙏

**Comment by xwu** [comment]
@swift-ci test

**Comment by YOCKOW** [comment]
I believe class declarations in `extension` such as [here](https://github.com/apple/swift-docc/blob/9808092bac6ddce8f9105c21fc678c857ef8b2e7/Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift#L457)  should not be warned, but:

```console
/Users/buildnode/jenkins/workspace/swift-PR-macos/branch-main/swift-docc/Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift:457:5: warning: non-'@objc' properties in extensions cannot be overridden; use 'public' instead
    open class Builder {
    ^~~~
    public
```

(Abridgment from https://ci.swift.org/job/swift-PR-macos/30379/consoleText )

**Comment by mininny** [comment]
> I believe class declarations in `extension` such as [here](https://github.com/apple/swift-docc/blob/9808092bac6ddce8f9105c21fc678c857ef8b2e7/Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift#L457) should not be warned, but:
> 
> ```
> /Users/buildnode/jenkins/workspace/swift-PR-macos/branch-main/swift-docc/Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift:457:5: warning: non-'@objc' properties in extensions cannot be overridden; use 'public' instead
>     open class Builder {
>     ^~~~
>     public
> ```
> 
> (Excerpt from https://ci.swift.org/job/swift-PR-macos/30379/consoleText )

Great catch! Thanks :) 
I added extra check for only emitting the warning for functions and variable declarations. 
And it seems like the tests are passing now as well

File: lib/Sema/TypeCheckAttr.cpp
Comment (xwu) [inline lib/Sema/TypeCheckAttr.cpp:877]: What about subscripts, initializers, etc.?

Comment (mininny) [reply lib/Sema/TypeCheckAttr.cpp:877]: Wouldn't they be included under `AbstractFunctionDecl` and `AbstractStorageDecl`? I was assuming we need to issue a warning for those cases as well; am I mistaken?

Comment (xwu) [reply lib/Sema/TypeCheckAttr.cpp:877]: Sorry, that’s my question to you, as I’m not sure; could it read more clearly explicitly to _exclude_ type declarations?

Comment (mininny) [reply lib/Sema/TypeCheckAttr.cpp]: I added test cases for subscripts, initializers and nominal types. Subscripts will work as open if they're @objc, but initializers won't work as initializers can't be open anyways. 
Also, I changed the expression to exclude `NominalTypeDecl` instead :)

File: test/attr/open_objc.swift
Comment (xwu) [inline test/attr/open_objc.swift:14]: Sweet!

I think this just needs a minor reworking so that the subscript is not referred to as a property; there are other diagnostics that have to deal with this so it'll be straightforward to adapt to your use case.

File: test/attr/open_objc.swift
Comment (xwu) [inline test/attr/open_objc.swift:21]: Nit: I know what you're getting at, but the comment is a little confusing, because despite the lack of your specific warning there is another expected error; I think you could just delete the comment and it'd be plenty clear.

File: include/swift/AST/DiagnosticsSema.def
Comment (LucianoPAlmeida) [inline include/swift/AST/DiagnosticsSema.def:1519]: To address @xwu comments about subscript being mentioned as property, we could use `DescriptiveDeclKind` here. 
```suggestion
      "non-'@objc' %0 in extensions cannot be overridden; use 'public' instead",
      (DescriptiveDeclKind))
```
And pass it as `VD->getDescriptiveKind()`. I think that would be a simpler way to address it :)

Comment (mininny) [reply include/swift/AST/DiagnosticsSema.def]: Thanks for the guide! Fixed it :)

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test

**Comment by xwu** [comment]
> > I believe class declarations in `extension` such as [here](https://github.com/apple/swift-docc/blob/9808092bac6ddce8f9105c21fc678c857ef8b2e7/Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift#L457) should not be warned, but:
> > ```
> > /Users/buildnode/jenkins/workspace/swift-PR-macos/branch-main/swift-docc/Sources/SwiftDocC/Indexing/Navigator/NavigatorIndex.swift:457:5: warning: non-'@objc' properties in extensions cannot be overridden; use 'public' instead
> >     open class Builder {
> >     ^~~~
> >     public
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (Excerpt from https://ci.swift.org/job/swift-PR-macos/30379/consoleText )
> 
> Great catch! Thanks :) I added extra check for only emitting the warning for functions and variable declarations. And it seems like the tests are passing now as well

Sorry, last thing: can we add a test for this?

**Review by xwu — APPROVED** [review]

**Comment by mininny** [comment]
Thanks! Can someone merge this, please? I don't have write access.

Come to think of it, what are the criteria to get write access? I have about 10 PRs merged to swift but am not sure [if it's non-trivial](https://www.swift.org/contributing/#commit-access) enough 🧐

**Pull request merged**
